### PR TITLE
SEO-577: 404 response on categories without members or media

### PR DESCRIPTION
**Pull request by saipetch**
SEO:
- 404 response on categories without members or media
- exclude category pages without members or media on SitemapXml

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 17 issues
* MAJOR 7 major
* MINOR 6 minor
* INFO 4 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SpecialSitemapXmlController.class.php#L191](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L191): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [SpecialSitemapXmlController.class.php#L220](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L30](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/Category.php#L30): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L191](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/Category.php#L191): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [Category.php#L256](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/Category.php#L256): Merge this if statement with the enclosing one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1066)
1. MAJOR [CategoryPage.php#L9](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/CategoryPage.php#L9): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [CategoryPage.php#L35](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/CategoryPage.php#L35): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)
1. MINOR [SpecialSitemapXmlController.class.php#L161](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L161): Immediately return this expression instead of assigning it to the temporary variable "$out". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. MINOR [SpecialSitemapXmlController.class.php#L186](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L186): Immediately return this expression instead of assigning it to the temporary variable "$out". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. MINOR [Category.php#L34](https://github.com/Wikia/app/blob/a21650e976088081b0cee9d9260bf334af2c559d/includes/Category.php#L34): Define a constant instead of duplicating this literal "cat_title" 4 times. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 17 issues
* MAJOR 7 major
* MINOR 6 minor
* INFO 4 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SpecialSitemapXmlController.class.php#L191](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L191): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [SpecialSitemapXmlController.class.php#L220](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L30](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/Category.php#L30): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L191](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/Category.php#L191): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [Category.php#L256](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/Category.php#L256): Merge this if statement with the enclosing one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1066)
1. MAJOR [CategoryPage.php#L9](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/CategoryPage.php#L9): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [CategoryPage.php#L35](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/CategoryPage.php#L35): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)
1. MINOR [SpecialSitemapXmlController.class.php#L161](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L161): Immediately return this expression instead of assigning it to the temporary variable "$out". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. MINOR [SpecialSitemapXmlController.class.php#L186](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L186): Immediately return this expression instead of assigning it to the temporary variable "$out". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. MINOR [Category.php#L34](https://github.com/Wikia/app/blob/9a4a040b4b7097903b1adf581be9b03916745f83/includes/Category.php#L34): Define a constant instead of duplicating this literal "cat_title" 4 times. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)

File: extensions/wikia/MercuryApi/MercuryApiController.class.php [hunk 1]
```diff
@@ -360,15 +360,17 @@ public function getPage() {
 					case NS_CATEGORY:
```
Comment (WikiaTech) [inline extensions/wikia/MercuryApi/MercuryApiController.class.php:360]: MAJOR Reduce this "switch/case" number of lines from 19 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)

File: extensions/wikia/MercuryApi/MercuryApiController.class.php [hunk 2]
```diff
@@ -360,15 +360,17 @@ public function getPage() {
...
-							$data['details'] = MercuryApiArticleHandler::getArticleDetails( $article );
-							$data['article'] = MercuryApiArticleHandler::getArticleJson( $this->request, $article );
-
-							// Remove namespace prefix from displayTitle, so it can be consistent with title
-							// Prefix shows only if page doesn't have {{DISPLAYTITLE:title} in it's markup
-							$data['article']['displayTitle'] = Title::newFromText($data['article']['displayTitle'])->getText();
-						} elseif ( !empty( $data['nsSpecificContent']['members']['sections'] ) ) {
-							$data['details'] = MercuryApiCategoryHandler::getCategoryMockedDetails( $title );
+						if ( !empty( $data['nsSpecificContent']['members']['sections'] ) ) {
+							if ( MercuryApiCategoryHandler::hasArticle( $this->request, $article ) ) {
```
Comment (WikiaTech) [inline extensions/wikia/MercuryApi/MercuryApiController.class.php:364]: MAJOR Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 27 issues
* MAJOR 12 major
* MINOR 11 minor
* INFO 4 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [MercuryApiController.class.php#L379](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/extensions/wikia/MercuryApi/MercuryApiController.class.php#L379): Reduce this "switch/case" number of lines from 16 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
1. MAJOR [MercuryApiController.class.php#L381](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/extensions/wikia/MercuryApi/MercuryApiController.class.php#L381): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
1. MAJOR [MercuryApiController.class.php#L517](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/extensions/wikia/MercuryApi/MercuryApiController.class.php#L517): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. MAJOR [SpecialSitemapXmlController.class.php#L191](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L191): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [SpecialSitemapXmlController.class.php#L220](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/extensions/wikia/SitemapXml/SpecialSitemapXmlController.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L30](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/includes/Category.php#L30): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [Category.php#L191](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/includes/Category.php#L191): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [Category.php#L256](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/includes/Category.php#L256): Merge this if statement with the enclosing one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1066)
1. MAJOR [CategoryPage.php#L9](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/includes/CategoryPage.php#L9): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. MAJOR [CategoryPage.php#L35](https://github.com/Wikia/app/blob/8a751feb5293d3383d2856b3b5f17660f71edcaa/includes/CategoryPage.php#L35): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 27 issues
- MAJOR 12 major
- MINOR 11 minor
- INFO 4 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. MAJOR [MercuryApiController.class.php#L379](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/extensions/wikia/MercuryApi/MercuryApiController.class.php#L379): Reduce this "switch/case" number of lines from 16 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
2. MAJOR [MercuryApiController.class.php#L381](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/extensions/wikia/MercuryApi/MercuryApiController.class.php#L381): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
3. MAJOR [MercuryApiController.class.php#L517](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/extensions/wikia/MercuryApi/MercuryApiController.class.php#L517): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
4. MAJOR [SitemapXmlController.class.php#L194](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L194): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
5. MAJOR [SitemapXmlController.class.php#L223](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L223): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
6. MAJOR [Category.php#L30](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/includes/Category.php#L30): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
7. MAJOR [Category.php#L191](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/includes/Category.php#L191): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
8. MAJOR [Category.php#L256](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/includes/Category.php#L256): Merge this if statement with the enclosing one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1066)
9. MAJOR [CategoryPage.php#L9](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/includes/CategoryPage.php#L9): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
10. MAJOR [CategoryPage.php#L35](https://github.com/Wikia/app/blob/26f9e616fb5237b35fce7a9e87cbf28729ba91b1/includes/CategoryPage.php#L35): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)

File: extensions/wikia/MercuryApi/MercuryApiController.class.php [hunk 3]
```diff
@@ -360,15 +360,17 @@ public function getPage() {
...
+								$data[ 'details' ] = MercuryApiArticleHandler::getArticleDetails( $article );
+								$data[ 'article' ] = MercuryApiArticleHandler::getArticleJson( $this->request, $article );
+
+								// Remove namespace prefix from displayTitle, so it can be consistent with title
+								// Prefix shows only if page doesn't have {{DISPLAYTITLE:title} in it's markup
+								$data[ 'article' ][ 'displayTitle' ] = Title::newFromText( $data[ 'article' ][ 'displayTitle' ] )->getText();
+							} else {
+								$data['details'] = MercuryApiCategoryHandler::getCategoryMockedDetails( $title );
+							}
```
Comment (gabrys) [inline extensions/wikia/MercuryApi/MercuryApiController.class.php:373]: Let's update the exception message to `Category has no members`

File: extensions/wikia/SitemapXml/SitemapXmlController.class.php [hunk 4]
```diff
@@ -147,6 +147,10 @@ private function generateSitemapPage( array $namespaces, $page ) {
 		}
 
 		foreach ( $this->model->getItems( $namespaces, $offset, $limit ) as $page ) {
+			if ( $page->page_namespace === NS_CATEGORY && !$this->isValidCategory( $page->page_title ) ) {
```
Comment (gabrys) [inline extensions/wikia/SitemapXml/SitemapXmlController.class.php:150]: How much overhead does it add to the NS_6-p\* sitemap URLs?

Comment (gabrys) [reply extensions/wikia/SitemapXml/SitemapXmlController.class.php:150]: I'd rather have a separate code to generate the link from the `categories` table directly.

File: includes/Category.php [hunk 5]
```diff
@@ -308,4 +308,24 @@ public function refreshCounts() {
...
+	/**
+	 * Check if the category has members (Wikia change)
+	 * @return bool
+	 */
+	public function hasMembers() {
+		$app = F::app();
+		$cacheKey = wfMemcKey( __METHOD__, sha1( $this->mName ) );
+		$data = $app->wg->Memc->get( $cacheKey );
+		if ( !is_numeric( $data ) ) {
+			if ( empty( $this->getPageCount() ) && empty( $this->getSubcatCount() ) && empty( $this->getFileCount() ) ) {
```
Comment (gabrys) [inline includes/Category.php:321]: `getPageCount()` returns the sum already.

File: includes/WikiCategoryPage.php [hunk 6]
```diff
@@ -12,18 +12,10 @@ class WikiCategoryPage extends WikiPage {
 	 * @return bool
 	 */
 	public function hasViewableContent() {
-		if ( parent::hasViewableContent() ) {
```
Comment (gabrys) [inline includes/WikiCategoryPage.php:15]: Why do we update this function?

File: includes/CategoryPage.php [hunk 7]
```diff
@@ -60,6 +60,11 @@ function view() {
 		parent::view();
 
 		if ( NS_CATEGORY == $title->getNamespace() ) {
+			/* Wikia change - begin */
```
Comment (gabrys) [inline includes/CategoryPage.php:63]: I'm not sure if this is a good example, but Wall uses `onArticleViewHeader` to set the status code. I'd rather not modify the MW files if not needed.

Long term idea: merge CategoryTree, CategoryPagination, CategoryGalleries and CategoryExhibitions into one WikiaCategory extension and use it to hold the hook code.

**Pull request closed without merging**